### PR TITLE
exemplar querying

### PR DESCRIPTION
**Pull request by cstyan**
Still needs a few more tests, but the implementation should be good for a first review.

Signed-off-by: Callum Styan <callumstyan@gmail.com>

### CODE REVIEW

File: pkg/distributor/query.go [hunk 1]
```diff
@@ -184,6 +211,81 @@ func (d *Distributor) queryIngesters(ctx context.Context, replicationSet ring.Re
...
+	i, j := 0, 0
+	for i < len(a) && j < len(b) {
+		if a[i].TimestampMs < b[j].TimestampMs {
+			result = append(result, a[i])
+			i++
+		} else if a[i].TimestampMs > b[j].TimestampMs {
+			result = append(result, b[j])
+			j++
+		} else {
+			result = append(result, a[i])
```
Comment (mdisibio) [inline pkg/distributor/query.go:227]: I'm not sure, but should this compare `Value` as well?   The Prometheus dedupe logic does so I think different exemplars with the same Ts but different Values are possible. https://github.com/prometheus/prometheus/blob/main/pkg/exemplar/exemplar.go#L49

File: pkg/distributor/query.go [hunk 2]
```diff
@@ -184,6 +211,81 @@ func (d *Distributor) queryIngesters(ctx context.Context, replicationSet ring.Re
 	return result, nil
 }
 
+// mergeExemplarSets merges and dedupes two sets of already sorted sample pairs.
+// Defined here instead of pkg/util to avoid a import cycle.
```
Comment (mdisibio) [inline pkg/distributor/query.go:215]: Might be useful to comment here that all exemplars are expected to be for the same series.

File: pkg/distributor/query.go [hunk 3]
```diff
@@ -53,6 +54,32 @@ func (d *Distributor) Query(ctx context.Context, from, to model.Time, matchers .
...
 
+func (d *Distributor) QueryExemplars(ctx context.Context, from, to model.Time, matchers ...[]*labels.Matcher) (*ingester_client.QueryResponse, error) {
+	var result *ingester_client.QueryResponse
+	err := instrument.CollectedRequest(ctx, "Distributor.QueryExemplars", d.queryDuration, instrument.ErrorCode, func(ctx context.Context) error {
+		req, err := ingester_client.ToExemplarQueryRequest(from, to, matchers...)
+		if err != nil {
+			return err
+		}
+
+		replicationSet, err := d.GetIngestersForQuery(ctx, false, nil)
```
Comment (mdisibio) [inline pkg/distributor/query.go:65]: I think it would be good to clarify how all matchers are being sent to all ingesters for the query, even though on ingest they are sharded by series (i.e. it is expected that an ingester will satisfy only a subset of the matchers and the results will be merged).  Maybe some comments, or a different name for the `exemplarQuery` parameter would be enough.

File: pkg/distributor/query.go [hunk 4]
```diff
@@ -184,6 +211,81 @@ func (d *Distributor) queryIngesters(ctx context.Context, replicationSet ring.Re
...
+		return nil, err
+	}
+
+	// Merge results from replication set.
+	var keys []string
+	exemplarResults := make(map[string]cortexpb.TimeSeries)
+	for _, result := range results {
+		r := result.(*ingester_client.QueryResponse)
+		for _, ts := range r.Timeseries {
+			lbls := cortexpb.FromLabelAdaptersToLabels(ts.Labels).String()
```
Comment (pracucci) [inline pkg/distributor/query.go:267]: [non blocking] I'm wondering performance impact of this. Have you tried writing a benchmark and comparing it with using `Hash()` for the key?  The `keys []string` could then be a `keys []struct{ labels ..., hash ...}`

**Review by pracucci — COMMENTED** [review]
Solid job @cstyan! Didn't find any issue. I left few minor comments I would be glad if you could take a look. Waiting for tests!

Could you also document the new API endpoints in `docs/api/_index.md`, please?

File: pkg/cortexpb/compat.go [hunk 5]
```diff
@@ -116,6 +117,30 @@ func FromMetricsToLabelAdapters(metric model.Metric) []LabelAdapter {
...
+			TimestampMs: e.Ts,
+		})
+	}
+	return result
+}
+
+func FromExemplarProtosToExemplars(es []Exemplar) []exemplar.Exemplar {
+	result := make([]exemplar.Exemplar, 0, len(es))
+	for _, e := range es {
+		result = append(result, exemplar.Exemplar{
```
[the remark this answers is not in the document]
Comment (cstyan) [reply pkg/cortexpb/compat.go:135]: Unless I'm missing something, I tried this out and didn't notice any difference in allocs or overall byes allocated.

Using `*(*exemplar.Exemplar)(unsafe.Pointer(&input))` as the exemplar and changing the `cortexpb.Exemplar` definition to
```
message Exemplar {
  // Exemplar labels, different than series labels
  repeated LabelPair labels = 1 [(gogoproto.nullable) = false, (gogoproto.customtype) = "LabelAdapter"];
  double value = 2;
  int64 timestamp_ms = 3;
  bool hasTs = 4;
}
```

Didn't make any difference to a benchmark for this function. Credit to @mdisibio for confirming we could make the change in the first place.

File: pkg/distributor/query.go (hunk 3 above)
Comment (pracucci) [reply pkg/distributor/query.go]: > I'm also now thinking of whether we should just split this up more, getting the replica set for each set of matchers and running the exemplar queries with those, and then merging the results.

I don't think it's worth the effort. Let me explain why.

Regardless of shuffle-sharding (which is about tenants sharding, not series), series can be sharded across ingesters into two ways:
1. By metric name only
2. By all series

When sharding by all series you have to query all ingesters, while when sharding by metric name all series for a given metric name are sharded always to the same ingester (+ replicas). That's why we can restrict the ingesters to query from when "shard by metric name only" is used.

However, the current implementation of "query only a restricted set of ingesters when shard by metric name only is enabled" is also buggy and doesn't take in account reshardings (eg. scale up), so it's not even safe to use out of the box with the blocks storage.

Long story short, I think it's fine (and desired) to always query all ingesters for the exemplars. When shuffle-sharding is enabled, it will only query ingesters belonging to the tenant's shard (and that's OK).

Comment (cstyan) [reply pkg/distributor/query.go]: @pracucci Thanks for the explanation! So in either scenario there's never 3 ingesters that have all of a given tenants data? A set of three would have a subset of their data based on the sharding type?

Comment (pracucci) [reply pkg/distributor/query.go]: Not sure I understand your question. Could you elaborate more your question, please?

File: docs/api/_index.md [hunk 6]
```diff
@@ -320,6 +321,21 @@ _For more information, please check out the Prometheus [range query](https://pro
...
+```
+GET,POST <prometheus-http-prefix>/api/v1/query_exemplars
+
+# Legacy
+GET,POST <legacy-http-prefix>/api/v1/query_exemplars
+```
+
+Prometheus-compatible exemplar query endpoint. When the request is sent through the query-frontend, the query will be accelerated by query-frontend (results caching and execution parallelisation).
+
+_For more information, please check out the Prometheus [range query](https://prometheus.io/docs/prometheus/latest/querying/api/#range-queries) documentation._
```
Comment (pracucci) [inline docs/api/_index.md:335]: Let's link to the right one.

```suggestion
_For more information, please check out the Prometheus [querying exemplars](https://prometheus.io/docs/prometheus/latest/querying/api/#querying-exemplars) documentation._
```

File: docs/api/_index.md [hunk 7]
```diff
@@ -320,6 +321,21 @@ _For more information, please check out the Prometheus [range query](https://pro
...
+### Exemplar query
+
+```
+GET,POST <prometheus-http-prefix>/api/v1/query_exemplars
+
+# Legacy
+GET,POST <legacy-http-prefix>/api/v1/query_exemplars
+```
+
+Prometheus-compatible exemplar query endpoint. When the request is sent through the query-frontend, the query will be accelerated by query-frontend (results caching and execution parallelisation).
```
Comment (pracucci) [inline docs/api/_index.md:333]: The query-frontend doesn't accelerate exemplars endpoint.

```suggestion
Prometheus-compatible exemplar query endpoint.
```

File: pkg/ingester/ingester_v2.go [hunk 8]
```diff
@@ -1031,6 +1035,56 @@ func (i *Ingester) v2Query(ctx context.Context, req *client.QueryRequest) (*clie
...
+	if err != nil {
+		return nil, err
+	}
+
+	from, through, matchers, err := client.FromExemplarQueryRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	i.metrics.queries.Inc()
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/ingester/ingester_v2.go:1049]: > I don't know if Cortex has a 'metrics label changes are breaking changes' policy.

The breaking changes policy doesn't cover metrics so no policy blocker here.

File: pkg/ingester/ingester_v2.go [hunk 9]
```diff
@@ -1031,6 +1035,56 @@ func (i *Ingester) v2Query(ctx context.Context, req *client.QueryRequest) (*clie
...
+
+	db := i.getTSDB(userID)
+	if db == nil {
+		return &client.ExemplarQueryResponse{}, nil
+	}
+
+	// Note that currently Prometheus' exemplar querier does nothing with a context that you pass it.
```
Comment (pracucci) [inline pkg/ingester/ingester_v2.go:1056]: [nit] I would remove this comment just because these kind of comments age very badly :) If ctx will be used at some point by Prometheus, we'll very likely forget to remove this comment, but this comment can still deceive the reader/contributor.

File: pkg/ingester/ingester_v2.go [hunk 10]
```diff
@@ -1031,6 +1035,56 @@ func (i *Ingester) v2Query(ctx context.Context, req *client.QueryRequest) (*clie
...
+			Labels:    cortexpb.FromLabelsToLabelAdapters(es.SeriesLabels),
+			Exemplars: cortexpb.FromExemplarsToExemplarProtos(es.Exemplars),
+		}
+
+		numExemplars += len(ts.Exemplars)
+		result.Timeseries = append(result.Timeseries, ts)
+	}
+
+	// TODO should we update this series metric again?
+	// i.metrics.queriedSeries.Observe(float64(len(result.Timeseries)))
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/ingester/ingester_v2.go:1082]: Thinking more about this, I believe we shouldn't. Because of this, I would remove this commented code.

File: pkg/ingester/metrics.go [hunk 11]
```diff
@@ -113,6 +114,12 @@ func newIngesterMetrics(r prometheus.Registerer, createMetricsConflictingWithTSD
 			// Could easily return 10m samples per query - 10*(8^(8-1)) = 20.9m.
 			Buckets: prometheus.ExponentialBuckets(10, 8, 8),
 		}),
+		queriedExemplars: promauto.With(r).NewHistogram(prometheus.HistogramOpts{
+			Name: "cortex_ingester_queried_exemplars",
+			Help: "The total number of exemplars returned from queries.",
+			// TODO: think about buckets, guessing here.
```
Comment (pracucci) [inline pkg/ingester/metrics.go:120]: To my understanding, max bucket is 1*(5^(5-1)) = 625. Maybe a bit low?

**Review by pracucci — COMMENTED** [review]
Thanks @cstyan ! The PR LGTM! I left few last minor comments. Could you also add at least a unit test to cover the querying part?

**Review by cstyan — COMMENTED** [review]
Sorry for the delay, on-call has been noisy this week.

Updated the ingester push test (which also calls the query function for samples) to call the query exemplars function and confirm exemplars that are stored properly (we only had 1 test case for storing exemplars in that test), hadn't pushed that yet so here it is.

Addressed a few other small comments + replied to some, mostly metrics related.

**Review by pracucci — APPROVED** [review]
LGTM, thanks! I left few final nits, all very small things.

> * [ENHANCEMENT] Blocks storage: support ingesting exemplars.  Enabled by setting new CLI flag `-blocks-storage.tsdb.max-exemplars=<n>` or config option `blocks_storage.tsdb.max_exemplars` to positive value. #4124

Could you update this CHANGELOG entry mentioning we support querying too + adding the PR number along side "#4124", please?

File: pkg/distributor/query.go (hunk 1 above)
Comment (cstyan) [reply pkg/distributor/query.go:229]: Thanks again for the detailed explanation.

File: pkg/api/handlers.go [hunk 12]
```diff
@@ -254,6 +256,8 @@ func NewQuerierHandler(
 	router.Path(path.Join(legacyPrefix, "/api/v1/read")).Methods("POST").Handler(legacyPromRouter)
 	router.Path(path.Join(legacyPrefix, "/api/v1/query")).Methods("GET", "POST").Handler(legacyPromRouter)
 	router.Path(path.Join(legacyPrefix, "/api/v1/query_range")).Methods("GET", "POST").Handler(legacyPromRouter)
+	// unclear to me whether we need to register here
```
[the remark this answers is not in the document]
Comment (cstyan) [reply pkg/api/handlers.go]: created: https://github.com/cortexproject/cortex/issues/4213

**Review by pracucci — DISMISSED** [review]
LGTM! Thanks!

**Comment by cstyan** [comment]
I found some issues while deploying/testing this that I will fix later this week.

File: pkg/distributor/query.go [hunk 13]
```diff
@@ -53,6 +54,32 @@ func (d *Distributor) Query(ctx context.Context, from, to model.Time, matchers .
 	return matrix, err
 }
 
+func (d *Distributor) QueryExemplars(ctx context.Context, from, to model.Time, matchers ...[]*labels.Matcher) (*ingester_client.QueryResponse, error) {
+	var result *ingester_client.QueryResponse
+	err := instrument.CollectedRequest(ctx, "Distributor.QueryExemplars", d.queryDuration, instrument.ErrorCode, func(ctx context.Context) error {
```
[the remark this answers is not in the document]
Comment (jtlisi) [reply pkg/distributor/query.go:59]: As a followup to this it may be worth updating the help text for queryDuration. Currently it's:

```
Time spent executing expression queries.
```

It may be worth changing it to

```
Time spent executing expression and exemplar queries.
```

File: pkg/distributor/query.go [hunk 14]
```diff
@@ -184,6 +212,82 @@ func (d *Distributor) queryIngesters(ctx context.Context, replicationSet ring.Re
...
+		r := result.(*ingester_client.QueryResponse)
+		for _, ts := range r.Timeseries {
+			lbls := cortexpb.FromLabelAdaptersToLabels(ts.Labels).String()
+			e, ok := exemplarResults[lbls]
+			if !ok {
+				exemplarResults[lbls] = ts
+				keys = append(keys, lbls)
+			}
+			// Merge in any missing values from another ingesters exemplars for this series.
+			e.Exemplars = mergeExemplarSets(e.Exemplars, ts.Exemplars)
```
Comment (jtlisi) [inline pkg/distributor/query.go:276]: **suggestion:** I may be missing something, but if I'm not mistaken, there is no need to call this function if there isn't an existing result, correct?

```suggestion
			} else {
				// Merge in any missing values from another ingesters exemplars for this series.
				e.Exemplars = mergeExemplarSets(e.Exemplars, ts.Exemplars)
			}
```

File: pkg/ingester/metrics.go [hunk 15]
```diff
@@ -113,6 +114,11 @@ func newIngesterMetrics(r prometheus.Registerer, createMetricsConflictingWithTSD
 			// Could easily return 10m samples per query - 10*(8^(8-1)) = 20.9m.
 			Buckets: prometheus.ExponentialBuckets(10, 8, 8),
 		}),
+		queriedExemplars: promauto.With(r).NewHistogram(prometheus.HistogramOpts{
+			Name:    "cortex_ingester_queried_exemplars",
+			Help:    "The total number of exemplars returned from queries.",
+			Buckets: prometheus.ExponentialBuckets(10, 5, 5),
```
Comment (jtlisi) [inline pkg/ingester/metrics.go:120]: **suggestion(optional):** I find comments like this helpful, may be worth adding.

```suggestion
			// A reasonable upper bound is around 6k - 10*(5^(5-1)) = 6250
			Buckets: prometheus.ExponentialBuckets(10, 5, 5),
```

File: pkg/distributor/query_test.go [hunk 16]
```diff
@@ -106,3 +109,52 @@ func TestMergeSamplesIntoFirstNilB(t *testing.T) {
 
 	require.Equal(t, b, a)
 }
+
+func TestMergeExemplarSets(t *testing.T) {
```
Comment (jtlisi) [inline pkg/distributor/query_test.go:113]: Clean test! 👍

**Review by jtlisi — APPROVED** [review]
LGTM

File: pkg/distributor/query.go (hunk 14 above)
Comment (cstyan) [reply pkg/distributor/query.go:276]: Yeah probably not, I don't think it's functionally any different. I guess there's the potential for slightly more memory being allocated than is really necessary without the else. Is this what you were thinking?

File: pkg/querier/distributor_queryable.go [hunk 17]
```diff
@@ -198,3 +200,48 @@ func (q *distributorQuerier) LabelNames() ([]string, storage.Warnings, error) {
...
+
+	if err != nil {
+		return nil, err
+	}
+
+	var e exemplar.QueryResult
+	ret := make([]exemplar.QueryResult, len(allResults.Timeseries))
+	for _, ts := range allResults.Timeseries {
+		// TODO: (callum) track down why we see empty exemplars slices here
+		// but not in the distributor/ingester code.
```
Comment (pracucci) [inline pkg/querier/distributor_queryable.go:238]: Can you elaborate on this? This looks something worth to investigate before merging this PR.

Comment (cstyan) [reply pkg/querier/distributor_queryable.go:238]: When I first deployed this branch to one of our Cortex clusters I noticed intermittent query failures in Grafana itself, and looking at the query response I saw a lot of results (they always seemed to be at the beginning of the overall `[]exemplar.QueryResult` slice) that had empty series labels and 0 len exemplar slices. The `if` condition right below was my quick fix. 

Yesterday I deployed these changes with additional debug logging, and this block was the only place I saw empty results. Which doesn't really make sense to me. This TODO is a note to me to look into this more. If there's a way to mock a query result from an Ingester replica set then I could try and reproduce in a test, didn't see anything obvious.

Comment (cstyan) [reply pkg/querier/distributor_queryable.go]: Fixed the issue in the latest commit(s).

**Comment by cstyan** [comment]
Let me know if I should update all the example docker compose setups the same as the one for local s3 as I've done in the latest commit.

**Review by pracucci — APPROVED** [review]
LGTM! Thanks for addressing all feedback 🙏 

> Let me know if I should update all the example docker compose setups the same as the one for local s3 as I've done in the latest commit.

Not strictly required (still a nice to have). Definitely not a blocker for this PR. We can add it anytime, whenever required.
